modules/apps/portal-store/portal-store-s3/src/main/java/com/liferay/portal/store/s3/S3FileCache.java

```diff
@@ -71,26 +71,36 @@ public class S3FileCache {
 	public void cleanUpCacheFiles() {
 		_calledCleanUpCacheFilesCount++;
 
-		if (_calledCleanUpCacheFilesCount <
-				_cacheDirCleanUpFrequency.intValue()) {
+		if ((_calledCleanUpCacheFilesCount <
+				_cacheDirCleanUpFrequency.intValue()) ||
+			_cacheCleanUpRunning) {
 
 			return;
 		}
 
 		synchronized (this) {
```

Comment:
@mbowerman I think the problem here is about the usage of `synchronized (this)`, as this is a pure waiting. I think we can avoid this problem by using `Lock.tryLock()` as it's "check-and-skip":

```
if (_lock.tryLock()) {
    try {
        ...
    }
    finally {
        _lock.unlock();
    }
}

...

private final Lock _lock = new ReentrantLock();
```

Reply:
Ah, perfect! I forgot about ReentrantLock. I like this solution better. Done.